### ISSUE TITLE
[css-color-4] How to handle infinite values in color functions?

### ISSUE THREAD

**Issue by cdoublev** [w3c/csswg-drafts#10507]
Unless otherwise specified, color channel values are not clamped, which applies to values produced by math functions, which can be infinite.

Infinite color channel values produce `NaN` in most conversion functions.

<details>
  <summary>Specifically</summary>

  - `color(xyz-d50) -> color(xyz-d65)`
  - `color(xyz-d65) -> rgb-linear()`
    - when `x` and `y` are the same infinite
    - when `x`/`y` and `z` are infinite
  - `color(xyz-d50) -> lab()`
    - when one or more channels are the same infinite
  - `color(xyz-d65) -> color(xyz-d50)`
    - when `x` and `y` are infinite
    - when `y` and `z` are infinite
    - when `x` and `z` are the same infinite
  - `color(xyz-d65) -> oklab()`
  - `oklab() -> color(xyz-d65)`
  - `rgb() -> hsl()`
    - when one or more channel is infinite
  - `hsl() -> rgb()`
    - when `s` is infinite and `l` is 0/100
    - when `l` is infinite
  - `hwb() -> rgb()`
    - when `w` is infinite
    - when `b` is -Infinity
  - `lab() -> color(xyz-d50)`
    - when `l` is infinite and `a` is the opposite infinite
    - when `l` and `b` are the same infinite
  - `lch() -> lab()`
  - `oklch() -> oklab()`
    - when `c` is infinite and `h` is 0
  - `rgb-linear() -> color(xyz-d65)`
    - when one or more channels are infinite and one or more are the opposite infinite

  This [CodePen](https://codepen.io/creativewave/pen/eYaKyVX) allows to play with conversion functions, which have comments for NaN cases.
</details>

In `colorjs.io@0.5.1`, `NaN` seems to be replaced by `none`, which seems incorrect:

```js
let color = new Color('hsl', [0, 0, Infinity])
color.to('srgb').toString(); // rgb(none none none)
```

Clarification on how to handle infinite values has already been asked in #8629, which was resolved with *no further change* because *if you put an infinite calculation into an rgba(), the behavior is well-defined: clamp to the allowed range*.

At least, this is not true for some channels of other color functions than `rgb()`, and for relative colors.

Following this [comment](https://github.com/w3c/csswg-drafts/issues/8629#issuecomment-1513228530), I tried to guess the color resulting from a color function specified with one infinite channel value. Since I am personally only interested by serialized values, I only did it for `hsl() -> rgb()` and `hwb() -> rgb()`, but it should presumably be done for all conversions producing `NaN`.

<details>
  <summary>When converting <code>hsl()</code> to <code>rgb()</code> and <code>saturation</code> is <code>+Infinity</code> or <code>-Infinity</code>:</summary>

  - when `l === 0`, red/green/blue are `0`
  - when `l < 0`, same as when `l > 0` but with inverted infinite signs
  - when `l > 0`:

  |             `h` |            `r` |            `g` |            `b`
  | --------------- | -------------- | -------------- | --------------
  |  `330 < h < 30` |    `+Infinity` |    `-Infinity` |    `-Infinity`
  |      `h === 30` |    `+Infinity` | depends on `l` |    `-Infinity`
  |   `30 < h < 90` |    `+Infinity` |    `+Infinity` |    `-Infinity`
  |      `h === 90` | depends on `l` |    `+Infinity` |    `-Infinity`
  |  `90 < h < 150` |    `-Infinity` |    `+Infinity` |    `-Infinity`
  |     `h === 150` |    `-Infinity` |    `+Infinity` | depends on `l`
  | `150 < h < 210` |    `-Infinity` |    `+Infinity` |    `+Infinity`
  |     `h === 210` |    `-Infinity` | depends on `l` |    `+Infinity`
  | `210 < h < 270` |    `-Infinity` |    `-Infinity` |    `+Infinity`
  |     `h === 270` | depends on `l` |    `-Infinity` |    `+Infinity`
  | `270 < h < 330` |    `+Infinity` |    `-Infinity` |    `+Infinity`
  |     `h === 330` |    `+Infinity` |    `-Infinity` | depends on `l`

  When converting `hsl()` to `rgb()` and `l === +Infinity`:

    - when `-100 <= saturation <= 100`, r, g, b, are `+Infinity`
    - when `saturation < -100`, same as `saturation > 100` but with inverted infinite signs
    - when `saturation > 100`, r/g/b are `+Infinity`/`-Infinity` depending on `saturation` and `h`
      - there are 6 intervals defined by  `h`
      - the exact `h` interval values depend on `saturation`
      - for each sibling intervals, one of r/g/b has a different sign

  When converting `hsl()` to `rgb()` and `l === -Infinity`:

    - when `-100 <= saturation <= 100`, r/g/b are `-Infinity`
    - when `saturation < -100` or `saturation > 100`, same as when `lightness === +Infinity` but with different intervals
</details>

Let's skip `hwb() -> rgb()` because it becomes more complicated when more than one infinite channel value.

For example, when `a === 0`, `x` tends towards `+Infinity` when `l` tends towards `+Infinity`, or towards `-Infinity` when `b` also tends towards `+Infinity`. Should `z` be `0`? Should some "precedence" between channel values be defined?

Note that channel values do not linearly tend towards an infinite value:

  - `rgb(from hsl(0  -99 calc(infinity)) r g b)` would resolve to `rgb(calc(infinity) calc(infinity) calc(infinity))`
  - `rgb(from hsl(0 -101 calc(infinity)) r g b)` would resolve to `rgb(calc(infinity) calc(-infinity) calc(-infinity))`

Am I missing something?

**Comment by svgeesus** [comment]
I agree that coercing values to infinity is going to cause no end of problems with color math; most color spaces are not designed to cope with such extreme values.

@tabatkins why are we doing this, again?

> Clarification on how to handle infinite values has already been asked in #8629, which was resolved with _no further change_ because _if you put an infinite calculation into an `rgba()`, the behavior is well-defined: clamp to the allowed range_.

For color spaces are clamped, yes. Most are not, including `color(srgb ...)` and `oklab()` and so on.

I just re-opened #8629 because the WG discussion apparently consisted of @tabatkins  asserting that it was fine (for `rgba()`) and thus fine everywhere else.

**Comment by romainmenke** [comment]
Not sure if this is applicable to browser implementations but I found that clamping to a really large value gives good results : https://github.com/csstools/postcss-plugins/blob/20ccaebde3de250e858adbbd3d519b01b7cf5304/packages/css-color-parser/src/functions/hsl-normalize-channel-values.ts#L111

Having a really large but finite value makes the color math work as expected.

**Comment by svgeesus** [comment]
∞ mod 360

**Comment by svgeesus** [comment]
The CSS Working Group just [discussed](https://www.w3.org/2025/08/19-css-color-minutes.html#1f28) `[css-color-4] How to handle infinite values in color functions?` and agreed to the following:

 * `RESOLUTION: clarify that color channels have an UA define limit for values, which affect infinity clamping`

<details><summary>The full IRC log of that discussion</summary>
TabAtkins: guillaume asked what do we do if you put infinity into one of the channels ? <br>
TabAtkins: when you do conversion, it will generally produce NaN<br>
TabAtkins: is it intended ?<br>
lea: how do we handle hue being infinite ?<br>
TabAtkins: I said that infinity is fine<br>
TabAtkins: it's simply a very big value<br>
TabAtkins: we should do the same for 1million or infinity<br>
TabAtkins: romain find clamping infinity to large value give good result for conversion<br>
ChrisL: it's wrong because any define number gives a actual result, infinity does not<br>
TabAtkins: for hue infinity it's 0<br>
TabAtkins: infinity shouldnt diverge from a very big number<br>
TabAtkins: if you mix something with big number hue, you still get a valid hue<br>
florian: its not define (different on browser/hardware) but it's still a value<br>
florian: for big value, we dont define how you clamp. its UA defined<br>
florian: in this specific case, we have coordinate space, and depending on how we clamp, it can make giant changes<br>
TabAtkins: just dont use big numbers<br>
ChrisL: we had another issue closed because you said it was fine on all colorspaces, while it was okay only on srgb<br>
TabAtkins: that's the same behavior for infinity in many places in CSS like length<br>
florian: the problem pre-exist in other places like width and height<br>
<TabAtkins> `width: calc(infinity * 1px)` will give you an arbitrary very large length, might not have any meaning to the author<br>
florian: even with infinity for both width and height you not sure having a square<br>
florian: there is just no guarantee whatsoever<br>
florian: dont expect something useful necessarily<br>
TabAtkins: its also true for any arbitrary large value<br>
TabAtkins: so are color so different that we need special behavior for infinity<br>
<ChrisL> So basically `calc(Infinity)` is legal, and please don't do that<br>
TabAtkins: or we just say author should not use infinity, whatever happen is undefined<br>
TabAtkins: the remaining question is when we do color conversion, do we want to specify in the spec that infinity and large value should be clamp to a specific large value<br>
ChrisL: I prefer not have that in the color spec, but in values & units<br>
TabAtkins: The spec says that if there are implementation limit, you should clamp infinity/large to this limit<br>
TabAtkins: we should make it clear that there is an implementation limit<br>
proposal: clarify that color channels have an UA define limit for values, which affect infinity clamping<br>
TabAtkins: you should get the same behavior for infinity and arbitrary large value<br>
RESOLUTION: clarify that color channels have an UA define limit for values, which affect infinity clamping<br>
</details>
